lib/engine-universal/src/code_memory.rs

```diff
@@ -2,217 +2,235 @@
+    map: *mut u8,
+
+    /// Mapping size
+    size: usize,
+
+    /// Size of the executable portion that comes first.
+    executable_size: usize,
+}
+
+impl CodeMemory {
+    fn create(size: usize) -> rustix::io::Result<CodeMemory> {
+        // TODO: figure out who does this.
+        assert!(size != 0);
+        // TODO: figure out why is this necessary.
+        let page_size = rustix::param::page_size();
+        let size = std::cmp::max(page_size, round_up(size, page_size));
+        let map = unsafe {
+            mm::mmap_anonymous(
+                std::ptr::null_mut(),
+                size,
+                ProtFlags::empty(),
+                MapFlags::SHARED,
+            )?
+        };
+        Ok(CodeMemory {
+            source_pool: None,
+            map: map.cast(),
+            executable_size: 0,
+            size,
+        })
+    }
+
+    /// Ensure this CodeMemory is at least of the requested size.
+    ///
+    /// This may invalidate any data previously written into the mapping.
+    pub fn resize(mut self, size: usize) -> rustix::io::Result<Self> {
+        if self.size < size {
+            let source_pool = self.source_pool.take();
+            unsafe {
+                mm::munmap(self.map.cast(), self.size)?;
+                std::mem::forget(self);
             }
+            Self::create(size).map(|mut m| {
+                m.source_pool = source_pool;
+                m
+            })
+        } else {
+            Ok(self)
         }
+    }
 
-        Ok((
-            function_result,
-            executable_section_result,
-            data_section_result,
-        ))
+    /// Write to this code memory from the beginning of the mapping.
+    ///
+    /// # Safety
+    ///
+    /// At the time this method is called, there should remain no dangling readable/executable
+    /// references to this `CodeMemory`, for the original code memory that those references point
+    /// to are invalidated as soon as this method is invoked.
+    pub unsafe fn writer(&mut self) -> CodeMemoryWriter<'_> {
+        mm::mprotect(
+            self.map.cast(),
+            self.size,
+            MprotectFlags::WRITE | MprotectFlags::READ,
+        )
+        .expect("TODO");
+        self.executable_size = 0;
+        CodeMemoryWriter {
+            memory: self,
+            offset: 0,
+        }
     }
 
-    /// Apply the page permissions.
-    pub fn publish(&mut self) {
-        if self.mmap.is_empty() || self.start_of_nonexecutable_pages == 0 {
-            return;
+    /// Publish the specified number of bytes as executable code.
+    pub unsafe fn publish(&mut self) {
+        mm::mprotect(
+            self.map.cast(),
+            self.executable_size,
+            MprotectFlags::EXEC | MprotectFlags::READ,
+        )
+        .expect("TODO");
+    }
+
+    /// Remap the offset into an absolute address within a read-execute mapping.
+    pub fn executable_address(&self, offset: usize) -> *const u8 {
```

Comment:
I don’t know, obtaining an address to contents of the map is not inherently an unsafe operation though, it is the `publish`/`writer` that end up prodding the dragons, so I feel like it is more natural to have unsafety there, rather than here. Analogous situation as with casting a mutable reference to a pointer vs dereferencing said pointer(s) to make a mutable reference out of it.

Reply:
Hmm so correct me if I'm misunderstanding, but the way I'm thinking is:
- CodeMemory is basically &mut [u8; _] on steroids
- `.publish()` / `.writer()` are helper functions that semantically toggle it between a RxCodeMemory and a RwCodeMemory, in a way that's not encoded in the type system because it would be cumbersome but still
- RxCodeMemory exposes a .executable_address() that returns a pointer valid only until a call to .publish() or .resize()
- RwCodeMemory exposes writer commands that are basically safe to call already, assuming no raw pointers are ever dereferenced. It also exposes a .writable_address(), which I'd love us to get rid of in favor of the safe functions, that returns a pointer valid only until a call to .writer() or .resize().

So my understanding is that the only thing that could go wrong is basically if a .*_address()-returned pointer is kept to outlive a CodeMemory-modifying call. Am I correct, or am I missing something?

If this is a correct summary, then I'd think the safety requirements should hang upon the people using the .*_address()-returned pointer, exactly in the same way as currently happens with reference/pointer casts as you mentioned: in my mind, the parallel is the following (pseudo-code):
```rust
let data = [0; 15]; let tab = &mut data; /**/ let code_memory = CodeMemory::new();
let ptr = tab.as_ptr(); /**/ let ptr = code_memory.writer_address();
unsafe { *ptr; /**/ *ptr; } // both are valid
std::mem::drop(data); /**/ code_memory.publish();
unsafe { *ptr; /**/ *ptr; } // both are invalid
```

We don't even need to make these functions unsafe, because they don't have to be, only usage of the returned pointers should be unsafe. But their documentation should clearly stipulate that the returned pointers are valid only until the next call to .writer() or .resize().

Does what I'm saying make sense?